### PR TITLE
Implement optional check that query has partition filter

### PR DESCRIPTION
**Pull request by JamesRTaylor**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/SystemSessionProperties.java [hunk 1]
```diff
@@ -125,6 +125,7 @@
     public static final String DYNAMIC_FILTERING_MAX_PER_DRIVER_ROW_COUNT = "dynamic_filtering_max_per_driver_row_count";
     public static final String DYNAMIC_FILTERING_MAX_PER_DRIVER_SIZE = "dynamic_filtering_max_per_driver_size";
     public static final String IGNORE_DOWNSTREAM_PREFERENCES = "ignore_downstream_preferences";
+    public static final String QUERY_PARTITION_FILTER_REQUIRED = "query_partition_filter_required";
```
Comment (JamesRTaylor) [inline presto-main/src/main/java/io/prestosql/SystemSessionProperties.java]: This could alternately be a hive specific session property to opt into this.

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                List<ColumnHandle> partitionColumns = getPartitionColumns(table);
+                if (!partitionColumns.isEmpty() && node.getEnforcedConstraint().isAll()) {
+                    String partitionColumnNames = partitionColumns.stream().map(n -> metadata.getColumnMetadata(session, table, n).getName()).collect(Collectors.joining(","));
+                    throw new VerifyException("Filter on partition column required: " + partitionColumnNames);
+                }
+                return null;
+            }
+
+            private List<ColumnHandle> getPartitionColumns(TableHandle handle)
+            {
```
Comment (JamesRTaylor) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java]: Not sure on the "right" way to find the partition columns of a table, but this seems to work (for hive anyway).

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                TableProperties properties = metadata.getTableProperties(session, handle);
+                if (properties.getTablePartitioning().isPresent()
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java]: "table partitioning" in this context is not necessarily related to hive partitions. It's an engine-level concept that describes whether data will be partitioned according to certain columns as it comes out of splits. Also, it's unrelated to whether a condition on a partition column is present in the query.

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import io.prestosql.sql.planner.plan.PlanVisitor;
+import io.prestosql.sql.planner.plan.TableScanNode;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static io.prestosql.SystemSessionProperties.isQueryPartitionFilterRequired;
+
+public class PartitionFilterChecker
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java]: This check should be specific to the Hive connector and be implemented within the connector. There's no concept of table partitioning in the engine.

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            private List<ColumnHandle> getPartitionColumns(TableHandle handle)
+            {
+                TableProperties properties = metadata.getTableProperties(session, handle);
+                if (properties.getTablePartitioning().isPresent()
+                        && !properties.getTablePartitioning().get().getPartitioningColumns().isEmpty()) {
+                    return properties.getTablePartitioning().get().getPartitioningColumns();
+                }
+                if (properties.getDiscretePredicates().isPresent()
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java]: There are a couple of issues with this check:
- `getDiscretePredicates` is not a well-supported API. We're trying to limit its scope of use and eventually get rid of it.
- As we continue to evolve the pushdown APIs (`applyXXX`), `getTableProperties` may not be able to produce a list of "discrete predicates" or table partitioning for a given "TableHandle". Historically, a TableHandle represented a base table, but that's no longer the case, as it potentially represents a base table with a number of operations applied to it (i.e., a form of virtual or derived table).

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java (hunk 4 above)
Comment (JamesRTaylor) [reply presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java]: Any suggestions on where this should be called int he Hive connector?

**Comment by JamesRTaylor** [comment]
Thanks for the feedback, @martint. I'll rework these in the next version of the patch. Couple of cases I've seen come up when testing against production queries that I'm curious if you have ideas on how to handle:
* Subquery that determines ds value:`SELECT ... WHERE ds IN (SELECT max(ds) from ...)`. Would there be a reasonable way to detect this case? The workaround is to force the user to tack on an `AND ds > 'some min date value'`
* Expression that determines partitions like this one: `WHERE (
                    CASE WHEN cardinality(split(ds, '-')) > 1
                    THEN cast(split(ds, '-')[2] as INT)
                    ELSE 0 END) % 11 = 0`

**Comment by electrum** [comment]
That form of subquery is actually not a partition predicate, since it’s evaluated at runtime as a join. It can be improved by dynamic filtering, but is still less efficient than a constant and the efficiency is dependent on execution timing. So you would be right to reject the query (based on your blocking goals).

The arbitrary expression actually does work since the engine passes it to the connector and the Hive connector uses it to filter partition names at planning. If your check is implemented by asserting that the selected partition list is smaller than the entire set of partitions, then this will work (and the user can’t cheat by passing a useless range filter).

**Comment by JamesRTaylor** [comment]
For the MAX subquery that turns into a join, would all partitions be scanned? It's a pattern I see used a fair amount, so if it's still reasonable, it'd be good to allow it.

Good to know about the expression being pushed down to the connector. My initial take is to allow the query to run if there's any filter or expression pushed down. As long as the user has considered a partition filter, I think that's good enough.

**Comment by electrum** [comment]
All partitions are scanned. The Hive connector produces pages with lazy blocks for data and a prefilled dictionary block with the partition value. The engine filters the partition columns and skips loading the data.

The pages need row counts. For columnar formats, it can get this from the file metadata sections. For text and the like, it has to read all the data.

So unfortunately, this is not a good pattern at all. Aggregation pushdown seems like it could solve this, but Hive allows empty partitions, and it’s not always cheap to tell if one is empty. Though empty partitions are uncommon, so we could probably make it work well enough (just check the min/max partition and hopefully it’s not empty).

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -156,6 +163,186 @@ protected QueryRunner createQueryRunner()
...
+                "create table partition_test(\n"
+                + "id integer,\n"
+                + "a varchar,\n"
+                + "b varchar,\n"
+                + "ds varchar)"
+                + "WITH (format='PARQUET', partitioned_by = ARRAY['ds'])");
+        assertUpdate(admin, "insert into partition_test(id,a,ds) values(1, 'a','a')", 1);
+        getQueryRunner().installPlugin(new TestingEventListenerPlugin(generatedEvents));
+        generatedEvents.initialize(2);
+        assertQueryFails(admin, "select id from partition_test where a = 'a'", "Filter required on tpch\\.partition_test for at least one partition column.*");
```
Comment (JamesRTaylor) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: The above query fails to generate a QueryCompletedEvent. Or maybe I'm not testing this correctly? We rely on the event to occur so that we can log the exception (and know which queries will fail this partition filter check before we enable this feature). Any ideas, @dain?

**Comment by JamesRTaylor** [comment]
Thanks for the previous review. I've moved this into the hive connector in my last commit. I'm not getting a QueryCompletedEvent if/when the partition filter check fails (see comment in PR). Any ideas on this?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 6 above)
Comment (JamesRTaylor) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Moved test and fix to https://github.com/prestosql/presto/pull/2544

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java [hunk 7]
```diff
@@ -172,6 +174,16 @@ public ConnectorSplitSource getSplits(
...
+        if (HiveSessionProperties.isQueryPartitionFilterRequired(session)) {
+            List<HiveColumnHandle> partitionColumns = hiveTable.getPartitionColumns();
+            if (!partitionColumns.isEmpty() && hiveTable.getEnforcedConstraint().isAll()) {
+                String partitionColumnNames = partitionColumns.stream().map(n -> n.getName()).collect(Collectors.joining(","));
+                throw new PrestoException(
+                    StandardErrorCode.QUERY_REJECTED,
+                    String.format("Filter required on %s.%s for at least one partition column: %s ", hiveTable.getSchemaName(), hiveTable.getTableName(), partitionColumnNames));
+            }
+        }
+
```
Comment (JamesRTaylor) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java]: Doing this check in getSplits() doesn't work well because EXPLAIN won't execute it (and thus will pass while query will fail). How about a new Connector method like this?

    /**
     * Allow the connector to validate the logical plan of the query to enable connector specific
    *  validation to occur.
    */
    default void validate(Plan logicalPlan) {}

Comment (martint) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java]: That could reasonable API to add. We've talked about adding explicit lifecycle steps before and after a query executes, but never got around to finalizing a design. Such a thing would help remove some "hacks" like `BeginTableWrite`.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java [hunk 8]
```diff
@@ -153,6 +157,17 @@ public HiveTableHandle withAnalyzeColumnNames(Set<String> analyzeColumnNames)
                 Optional.of(analyzeColumnNames));
     }
 
+    @Override
+    public void validateScan(ConnectorSession session)
+    {
+        if (HiveSessionProperties.isQueryPartitionFilterRequired(session) && !partitionColumns.isEmpty() && getEnforcedConstraint().isAll()) {
+            String partitionColumnNames = partitionColumns.stream().map(n -> n.getName()).collect(Collectors.joining(","));
```
Comment (suryanshagnihotri) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Will it allow queries with predicate expression on partition column?

Comment (JamesRTaylor) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Can you give me an example? If the query ends up with an enforced constraint, it'll be allowed.

**Comment by JamesRTaylor** [comment]
Anyone know how to rerun the build (short of pushing a bogus commit)?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java (hunk 8 above)
Comment (suryanshagnihotri) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Suppose there is a table 'T' with a partitioned column 'C' of type 'varchar' and you fired
```select * from T where CAST(C as int) = 10```, then I think this will throw an exception but it shouldn't, right?

Comment (JamesRTaylor) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Yes, that'd be flagged as an error (assuming the partition filter enforcement feature is enabled). I suspect that query would execute more efficiently written as `select * from T where C = '10'` which is the kind of thing this feature is supposed to catch.

Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Partition pruning happens in both cases, in case of `C = '10'` the predicate can be pushed down hive metastore as well to improve planning time. In case of `CAST(C as int) = 10` Presto will still run an expression interpreter to prune partitions after the hive metastore calls. So I think we'd want to allow such cases since partition pruning is performed and some extra CASTs or date/time conversion functions are often present in predicates on partition columns in queries generated by BI tools.

Comment (JamesRTaylor) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Makes sense - thanks for the suggestion. I've pushed a commit so that this case passes the partition filter check.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java [hunk 9]
```diff
@@ -153,6 +189,19 @@ public HiveTableHandle withAnalyzeColumnNames(Set<String> analyzeColumnNames)
                 Optional.of(analyzeColumnNames));
     }
 
+    @Override
+    public void validateScan(ConnectorSession session)
+    {
+        if (HiveSessionProperties.isQueryPartitionFilterRequired(session) && !partitionColumns.isEmpty()
+                && getEnforcedConstraint().isAll()
+                && (!getPartitionConstraint().isPresent() || !getPartitionConstraint().get().predicate().isPresent())) {
```
Comment (suryanshagnihotri) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: could you please explain how `!getPartitionConstraint().get().predicate().isPresent()` is helping us here.
I see we are propagating constraints from `pushFilterIntoTableScan` method where the constraint is initialised like `constraint = new Constraint(newDomain);` making predicate as `Optional.empty()` since `pruneWithPredicateExpression` is `false` so shouldn't `getPartitionConstraint().get().predicate().isPresent()` will always be `false`?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 10]
```diff
@@ -166,6 +166,211 @@ protected TestTable createTableWithDefaultColumns()
...
+                + "WITH (format='PARQUET')");
+        assertUpdate(admin, "insert into partition_test9(id,a,ds) values(1, 'a','a')", 1);
+        assertUpdate(admin, "insert into partition_test10(id,a,ds) values(1, 'a','a')", 1);
+        assertQuery(admin, "select a.id, b.id from partition_test9 a join partition_test10 b on (a.id = b.id) where a.ds = 'a'", "SELECT 1, 1");
+        assertUpdate(admin, "DROP TABLE partition_test9");
+        assertUpdate(admin, "DROP TABLE partition_test10");
+    }
+
+    @Test
+    public void testPartitionPredicateAllowed() throws Exception
```
Comment (JamesRTaylor) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Added test for dynamic partition pruning passing with query_partition_filter_required set to true.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java (hunk 9 above)
Comment (JamesRTaylor) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: The presence of the predicate on the partition constraint that was used in applyPartitionResult is an indication that there's dynamic partition pruning occurring. The constraint passed through will have a predicate in cases in which this occurs. Did you see the test I added [here](https://github.com/prestosql/presto/pull/2334/files#r375998776)?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java [hunk 11]
```diff
@@ -210,7 +210,7 @@ public HivePartitionResult getPartitions(ConnectorTableHandle tableHandle, List<
         return partitionList.build();
     }
 
-    public HiveTableHandle applyPartitionResult(HiveTableHandle handle, HivePartitionResult partitions)
+    public HiveTableHandle applyPartitionResult(HiveTableHandle handle, HivePartitionResult partitions, Optional<Constraint> partitionConstraint)
```
[the remark this answers is not in the document]
Comment (JamesRTaylor) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java]: Yes, we're checking whether the constraint has a predicate in validateScan. The change I've made to HiveTableHandle is to remember the dynamic partition pruning constraint we applied so we can test for it in validateScan.

**Comment by JamesRTaylor** [comment]
This PR is ready for review if you have a few spare cycles, @martint or @electrum.

**Comment by JamesRTaylor** [comment]
I've noticed one case that I can't explain. The following query passes the partition filter check, though it's not filtering on any partition column (assume there's a DS column that is the one being partitioned on). The io explain plan shows a constraint of the ds column being between the oldest to newest partition.

    SELECT count(*) from my_table 
    WHERE CAST(created_on AS date) >= CURRENT_DATE - INTERVAL '30' day;

Oddly enough, if I remove the unnecessary CAST, it no longer passes:

    SELECT count(*) from my_table 
    WHERE created_on >= CURRENT_DATE - INTERVAL '30' day;

Any ideas on how the first query can be detected as failing the partition filter check?

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorTableHandle.java [hunk 12]
```diff
@@ -15,4 +15,8 @@
 
 public interface ConnectorTableHandle
 {
+    /**
+     * Allow the connector to reject the table scan produced by the planner.
+     */
+    default void validateScan(ConnectorSession session) {}
```
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorTableHandle.java]: This should be implemented as a ConnectorMetadata API. Handles are meant to be data objects, not service APIs (think of them as some kind of opaque ID)

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+import io.prestosql.execution.warnings.WarningCollector;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.sql.planner.SimplePlanVisitor;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+
+public class PartitionFilterChecker
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/PartitionFilterChecker.java]: This is should be renamed to something like "ValidateTableScans" or "TableScanValidator".

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java [hunk 14]
```diff
@@ -176,7 +178,13 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
...
                             // which would be expensive to evaluate in the call to isCandidate below.
                             domainTranslator.toPredicate(newDomain.simplify().transform(assignments::get))));
-            constraint = new Constraint(newDomain, evaluator::isCandidate);
+            Optional<DiscretePredicates> discretePredicates = metadata.getTableProperties(session, node.getTable()).getDiscretePredicates();
+            if (discretePredicates.isPresent() && Collections.disjoint(discretePredicates.get().getColumns(), evaluator.getArguments())) {
+                constraint = new Constraint(newDomain);
+            }
+            else {
+                constraint = new Constraint(newDomain, evaluator::isCandidate);
+            }
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: Why is this needed? If possible, we want to avoid using the `getDiscretePredicates()` API since we're trying to get rid of it.

Comment (JamesRTaylor) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: Something like this is needed to prevent a predicate from being created on the Constraint and causing the partition filter check from incorrectly passing. See [these](https://github.com/prestosql/presto/pull/2334/files#diff-f735a7bf0d40c13be5c34d9a25110a08R366-R414) tests for examples. The first test is an example of a query that should pass the check (since it's pruning partitions) while the second shouldn't pass (since it's not even filtering on a partition column). The current logic in Presto creates a predicate that always returns true in the case of the filter not containing a reference to any partition columns (for Hive) [here](https://github.com/prestosql/presto/pull/2334/files#diff-64a95c3a25c48aeecc5f744e3e45b7b8L298-L300). The check I implemented [here](https://github.com/prestosql/presto/pull/2334/files#diff-7e85fbe1d8919e64b7cf7bcb615e12f4R197) differentiates these based on the existence of a predicate on the constraint.

Maybe this could be a method on ConnectorMetadata instead?

**Comment by tooptoop4** [comment]
do these give error or are allowed as there is a partition filter (although meaningless)?
select * from table where partition_col is not null
select * from table where partition_col = partition_col
select * from table where partition_col > ''
select * from table where partition_col in (select partition_col from table)
select * from table where partition_col >= (select min(partition_col ) from table)
select * from table where partition_col <= (select max(partition_col ) from table)
select * from table where partition_col in ('a','b','c')..etc where ALL partition values are listed

**Comment by JamesRTaylor** [comment]
@tooptoop4 - see tests I added in TestHiveIntegrationSmokeTest which cover some of these cases. Some of these (like the min/max subquery) were discussed here already too. In general, we just want to make sure that the user has thought about the partition column to catch cases in which they don't know or forgot about filtering on it. We're not trying to make a full proof mechanism (which is not actually feasible).

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java (hunk 14 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: Yeah, this is problematic because it might break connectors that can leverage the LayoutEvaluator::isCandidate logic but do not expose DiscretePredicates (which are a hack that only exists for legacy reasons).

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java [hunk 15]
```diff
@@ -223,7 +223,8 @@ public HiveTableHandle applyPartitionResult(HiveTableHandle handle, HivePartitio
                 partitions.getBucketHandle(),
                 partitions.getBucketFilter(),
                 handle.getAnalyzePartitionValues(),
-                handle.getAnalyzeColumnNames());
+                handle.getAnalyzeColumnNames(),
+                partitionConstraint);
```
Comment (martint) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java]: The Constraint object is only meant to be used by the immediate callees and not intended to be stored or used once the call is complete (it may hold on to unspecified resources under the covers -- e.g., compiled code, etc., or depend on state that is not available once the call completes). Therefore, it can't be stored in the handle for use later on.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java (hunk 14 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java]: I think what we could do is extend Constraint to include the list of columns included in the filter expression. With that, the connector should be able to keep track of which columns are being considered for filtering and fail the query during the call to `validateScan` if at least one partition column doesn't appear.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 16]
```diff
@@ -1857,6 +1860,24 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
...
+    @Override
+    public void validateScan(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        HiveTableHandle handle = (HiveTableHandle) tableHandle;
+        if (HiveSessionProperties.isQueryPartitionFilterRequired(session) && handle.getEnforcedConstraint().isAll()) {
+            List<HiveColumnHandle> partitionColumns = handle.getPartitionColumns();
+            if (!partitionColumns.isEmpty()) {
+                Optional<Set<ColumnHandle>> referencedColumns = handle.getConstraintColumns();
+                if (!referencedColumns.isPresent() || Collections.disjoint(referencedColumns.get(), partitionColumns)) {
+                    String partitionColumnNames = partitionColumns.stream().map(n -> n.getName()).collect(Collectors.joining(","));
```
Comment (martint) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Use method reference: `HiveColumnHandle::getName`. Also, format across multiple lines for readability.

```suggestion
                    String partitionColumnNames = partitionColumns.stream()
                            .map(HiveColumnHandle::getName)
                            .collect(Collectors.joining(","));
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java [hunk 17]
```diff
@@ -107,6 +108,35 @@ public HiveTableHandle(
...
+                compactEffectivePredicate,
+                enforcedConstraint,
+                bucketHandle,
+                bucketFilter,
+                analyzePartitionValues,
+                analyzeColumnNames,
+                Optional.empty());
+    }
+
+    public HiveTableHandle(
```
Comment (martint) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Instead of a new constructor, just modify the main one and adjust the callsites that invoke it. There are only a handful. 

BTW, because of how it's done here, there's a theoretical bug (or rather, inconsistency) due to `HiveMetadata.makeCompatiblePartitioning()` wanting to create a copy of the handle by invoking the constructor that doesn't take `constraintColumns`. When that gets called, any value that was set for that field will get lost. It doesn't matter in practice because that happens after the call to `validateScan`, but it's just good hygiene to make it be as consistent as possible.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 18]
```diff
@@ -1846,7 +1849,7 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
 
         HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, new HiveIdentity(session), handle, constraint);
-        HiveTableHandle newHandle = partitionManager.applyPartitionResult(handle, partitionResult);
+        HiveTableHandle newHandle = partitionManager.applyPartitionResult(handle, partitionResult, constraint.getColumns());
```
Comment (martint) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: `applyFilter` may be called multiple times throughout the optimization process, so this needs to combine the existing columns references in the handle, if any, with the new ones.

You should be able to reproduce this with a test that does something like:

```
SELECT * FROM (
    SELECT * FROM table
    WHERE partition_column = 'xxx'
)
WHERE non_partition_column = 'yyy'
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 19]
```diff
@@ -166,6 +166,252 @@ protected TestTable createTableWithDefaultColumns()
...
+        String query = "select id from partition_test3 where ds = 'a'";
+        assertQuery(admin, query, "select 1");
+        computeActual(admin, "explain " + query);
+        assertUpdate(admin, "DROP TABLE partition_test3");
+    }
+
+    @Test
+    public void testPartitionFilterIncluded2()
+    {
+        Session admin = Session.builder(getQueryRunner().getDefaultSession())
```
Comment (martint) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Instead of `admin`, rename this variable to `session` (same in the other tests)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 18 above)
Comment (JamesRTaylor) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Good catch. I added a test like this which was failing. I found that I didn't need to combine the existing column references with the new ones as the LayoutConstraintEvaluator appears to already combine the expressions so we already get the combined set of columns from the constraint. Instead, I had to add a condition to the if statement below to compare handle.getConstraintColumns() and newHandle.getConstraintColumns(). Otherwise, you lose the constraint column information when Optional.empty() is returned. .

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 20]
```diff
@@ -1846,7 +1849,7 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
 
         HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, new HiveIdentity(session), handle, constraint);
```
Comment (JamesRTaylor) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Do you think it'd be better to store the set of referenced columns in HivePartitionResult and not add a new argument to partitionManager.applyPartitionResult method?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 18 above)
Comment (martint) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: There's no guarantee the predicates will all be applied in one shot. Actually, the case I was thinking of is more complicated to reproduce. It can happen when there's a join and each side is progressively narrowed down as the predicate pushdown logic is applied by predicates that are inferred from each side on each iteration. E.g. the left side produces a list of partitions that constrains the right side, which produces tighter predicate that further constrains the left side.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 20 above)
Comment (martint) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Actually, both options kind of suck. The way I'd do it is by introducing a builder class and do something like:

```
HiveTableHandle newHandle = HiveTableHandle.builderFrom(handle)
    .withColumns(constraint.getColumns())
    .build()
```

The same mechanism would be used in `applyPartitionResult` to copy the original handle and override a few fields.

But don't worry about it. It's something we can address later. Let's keep the argument in `applyPartitionResult` for now.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java [hunk 21]
```diff
@@ -223,7 +226,9 @@ public HiveTableHandle applyPartitionResult(HiveTableHandle handle, HivePartitio
                 partitions.getBucketHandle(),
                 partitions.getBucketFilter(),
                 handle.getAnalyzePartitionValues(),
-                handle.getAnalyzeColumnNames());
+                handle.getAnalyzeColumnNames(),
+                Optionals.combine(handle.getConstraintColumns(), columns,
+                        (oldColumns, newColumns) -> Sets.union(oldColumns, newColumns)));
```
Comment (JamesRTaylor) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java]: Combine old and new column references here. I copied io.prestosql.util.Optionals into io.prestosql.plugin.hive.util as I didn't see an obvious module to move it to.

**Comment by tooptoop4** [comment]
@JamesRTaylor if the config flag is on but the table being queried is not a partitioned table does it allow? i couldn't see a test for that
